File: erlang/seminars/barber/customer.erl
-module(customer).

-export([start/2]).

-define(Walk, 1000).

start(Name, Shop) ->
    {ok, spawn_link(fun() -> init(Name, Shop) end)}.
		        
init(Name, Shop) ->
    io:format("~s - going for a hair cut~n", [Name]),
    customer(Name, Shop).

customer(Name, Shop) ->
    io:format("~s - enter the shop~n", [Name]),
    Shop ! {hello, self()},
    receive
	please_wait ->
	    io:format("~s - waiting~n", [Name]),
	    receive 
		have_a_seat ->
		    io:format("~s - taking a seat~n", [Name]), 
		    receive 
			howz_that ->
			    io:format("~s - thank you~n", [Name])
		    end
	    end;
	sorry ->
	    io:format("~s - taking a walk in the sun~n", [Name]),
	    timer:sleep(?Walk),
	    customer(Name, Shop)
    end.

File: exercises/barber/src/customer.ex
defmodule Customer do
  
  @walk 1000

  # Start a customer process.
  def start(name, shop) do
    {:ok, spawn_link(fn -> init(name, shop) end)}
  end

  defp init(name, shop) do
    IO.puts("#{name} - going for a haircut")
    customer(name, shop)
  end

  # When a customer enters the shop three different
  # scenarios can happen:
  # 1) the barber is free and will give the haircut
  # 2) the barber is busy but there is space in the waiting
  #    queue; the customer is asked to wait
  # 3) the barber is busy and there is no space in the
  #    waiting queue; the customer is invited to take a
  #    walk and return later
  defp customer(name, shop) do
    IO.puts("#{name} - enter the shop")
    send(shop, {:hello, self()})

    receive do
      :please_wait ->
        IO.puts("#{name} - waiting")

        receive do
          :have_a_seat ->
            IO.puts("#{name} - taking a seat")

            receive do
              :howz_that ->
                IO.puts("#{name} - thank you")
            end
        end

      :sorry ->
        IO.puts("#{name} - taking a walk in the sun")
        :timer.sleep(@walk)
        customer(name, shop)
    end
  end
end
